cognite/client/_api/spaces.py

```diff
@@ -0,0 +1,187 @@
+
+        """
+        identifiers = DataModelingIdentifierSequence.load_spaces(spaces=spaces)
+        return self._retrieve_multiple(list_cls=SpaceList, resource_cls=Space, identifiers=identifiers)
+
+    def list(
+        self,
+        limit: int = LIST_LIMIT_DEFAULT,
+    ) -> SpaceList:
+        """`List spaces <https://docs.cognite.com/api/v1/#tag/Spaces/operation/listSpacesV3>`_
+
+        Args:
+            limit (int, optional): Maximum number of spaces to return. Defaults to 25. Set to -1, float("inf") or None
+                to return all items.
+
+        Returns:
+            SpaceList: List of requested spaces
+
+        Examples:
+
+            List spaces and filter on max start time::
+
+                >>> from cognite.client import CogniteClient
+                >>> c = CogniteClient()
+                >>> space_list = c.data_modeling.spaces.list(limit=5)
+
+            Iterate over spaces::
+
+                >>> from cognite.client import CogniteClient
+                >>> c = CogniteClient()
+                >>> for space in c.data_modeling.spaces:
+                ...     space # do something with the space
+
+            Iterate over chunks of spaces to reduce memory load::
+
+                >>> from cognite.client import CogniteClient
+                >>> c = CogniteClient()
+                >>> for space_list in c.data_modeling.spaces(chunk_size=2500):
+                ...     space_list # do something with the spaces
+        """
+        return self._list(
+            list_cls=SpaceList,
+            resource_cls=Space,
+            method="GET",
+            limit=limit,
+        )
+
+    @overload
+    def apply(self, space: Sequence[Space]) -> SpaceList:
+        ...
+
+    @overload
+    def apply(self, space: Space) -> Space:
+        ...
+
+    def apply(self, space: Space | Sequence[Space]) -> Space | SpaceList:
+        """`Create or patch one or more spaces. <https://docs.cognite.com/api/v1/#tag/Spaces/operation/ApplySpaces>`_
+
+        Args:
+            space (space: Space | Sequence[Space]): Space or spaces of spacesda to create or update.
+
+        Returns:
+            Space | SpaceList: Created space(s)
+
+        Examples:
+
+            Create new spacesda::
+
+                >>> from cognite.client import CogniteClient
+                >>> from cognite.client.data_classes import Space
+                >>> c = CogniteClient()
+                >>> spaces = [Space(space="mySpace", description="My first space", name="My Space"),
+                ... Space(space="myOtherSpace", description="My second space", name="My Other Space")]
+                >>> res = c.data_modeling.spaces.create(spaces)
+        """
+        return self._create_multiple(list_cls=SpaceList, resource_cls=Space, items=space)
+
+    def delete(
+        self,
+        space: str | Sequence[str],
+    ) -> SpaceList | None:
```

Comment:
The DM APIs don't return the full resource when deleted. Just the identifiers. So it should be `Sequence[str] | None`.

Also, I guess it will be `str | None` when input is `str` and `Sequence[str]` when input is `Sequence[str]`? Let's either add overloads to express that or always make it return `Sequence[str]`.

Reply:
I think it should be `str | list[str] | None` as when we return we will know that it is a list, and do not need to make the generic `Sequence`.

Reply:
Yeah, agreed.